### ISSUE TITLE
ReScience challenge: resurrecting a 30-year-old Turbo Pascal program for generating Maucha ionic diagrams

### ISSUE THREAD

**Issue by MikeSilberbauer** [ReScience/submissions#25]
**Original article:**
Silberbauer, M.J. and King, J.M., 1991. Geographical Trends in the Water Chemistry of Wetlands in the South-Western Cape Province, South Africa. Southern African Journal of Aquatic Sciences, 17 (1/2), 82–88.
**PDF URL:**
[Silberbauer-2020.pdf](https://github.com/ReScience/submissions/files/4522154/ReScience_C_Silberbauer_2020_5d.pdf)
**Metadata URL:**
http://www.riv.co.za/wv/data/Silberbauer-2020.yaml
**Code URL:**
https://archive.softwareheritage.org/browse/origin/https://github.com/MikeSilberbauer/maucha-ionic-diagram.git/
**Scientific domain:**
Aquatic Science
**Programming language:**
Turbo Pascal
**Suggested editor:**

**Comment by rougier** [comment]
@MikeSilberbauer Thansk for your submission! Before we can proceed with the actual review, would it be possible to use the ReScience template for your submission? Also, After a quick read of your submission, it does not seem to have any figure showing actual results (both old and new) but only a link to some (really nice) vintage screenshots. I think you should add them to your submission (we have no limit page) since the provided link may dissappear in the future. 

@karthik @ThomasA Could you edit this submission for the Ten Years Reproducibility Challenge (only 1 reviewer needed) ?

**Comment by MikeSilberbauer** [comment]
Sorry about the delay in reacting. I have not used github much and I did not notice that in the settings my primary email for notifications was my old work address, which was blocked when I retired at the end of January.
Thank you for the positive response - I will try to work out how to use the template.

**Comment by MikeSilberbauer** [comment]
@rougier 
[ReScience_C_Silberbauer_2020_5d.pdf](https://github.com/ReScience/submissions/files/4522154/ReScience_C_Silberbauer_2020_5d.pdf)
With much assistance from Overleaf, I have converted the draft to LaTex and then PDF. I have not yet mastered the citation functions, but the document should now be in a state where it is reviewable.

**Comment by rougier** [comment]
@khinsen @pdebuyl Can you handle this submission for the Ten Years Reproducibility Challenge (only 1 reviewer needed)?

**Comment by rougier** [comment]
@ReScience/associate-editors (kind of urgent) Can any of you handle this submission for the Ten Years Reproducibility Challenge (only 1 reviewer needed)?

**Comment by oliviaguest** [comment]
I can try, yes, but it's far out of my area. Maybe if somebody can suggest reviewers?

**Comment by rougier** [comment]
Thanks. And in fact, I can probably review it. I've started reading the paper and it is quite fun!

**Comment by oliviaguest** [comment]
Fantastic! 💯

**Comment by rougier** [comment]
Here is my review:

The paper is both fun and instructive to read. I like the author's sense of humor.
I think it needs only a few minor modifications that I list here in no particular order:

* Please use `section*` instead of `subsection` for top levels section (* is for removing the section / subsection number)
* Since the source is not very long, you should include the source of the program in the article and give as well the link to software heritage inside the article. 
* I would be curious to know a bit more about where the source were located on your hard-drive because they are almost 30 years old. I'm probably a bit younger but I'm sure I've already lost most of the files I've created only 20 years ago.
* You could comment a bit on the source. Do you still understand what you were coding at that time ? For example, what are these "magic numbers" on lines 77-82 ?
* In your repository, you should add the CHEMDATA.DAT file, else people won't be able to run it.
* How did you find the original price of the turbo pascal compiler ?
* I would be nice to have a conclusion with your overall feeling about this successful reproduction
* Metadata will need some modification
* In the README of https://github.com/MikeSilberbauer/maucha-ionic-diagram, you could probably add a few more explanation and a link to this review and to the article once published.
* Note that I did not try to run the code.

This submission of the oldest reproduction for the Ten Years Reproducibility Challenge (almost 30 years) and it seems that the author did not encounter any problem for running the code. That's funny because some of the more recent submissions had some problems.

**Comment by MikeSilberbauer** [comment]
[ReScience_C_Silberbauer_2020 (7).pdf](https://github.com/ReScience/submissions/files/4832693/ReScience_C_Silberbauer_2020.7.pdf)

Thank you for your constructive comments. I have tried to deal with them in the updated PDF document. I have not yet edited the metadata, as I am not sure about the Software Heritage link.

**Comment by rougier** [comment]
For the software heritage link, it's mostly a matter of visiting https://archive.softwareheritage.org/save/ and you have to enter the address of your github repository and you're done. You should then obtain your swid

**Comment by MikeSilberbauer** [comment]
**Original article**:
Silberbauer, M.J. and King, J.M., 1991. Geographical Trends in the Water Chemistry of Wetlands in the South-Western Cape Province, South Africa. Southern African Journal of Aquatic Sciences, 17 (1/2), 82–88.
**PDF URL**:
Silberbauer-2020.pdf
**Metadata URL**:
http://www.riv.co.za/wv/data/Silberbauer-2020.yaml
**Code URL**:
https://github.com/MikeSilberbauer/maucha-ionic-diagram
**Scientific domain**:
Aquatic Science
**Programming language**:
Turbo Pascal
**Suggested editor**:
Nicolas P. Rougier

**Comment by rougier** [comment]
@MikeSilberbauer Not sure about the meaning of your post actually.

**Comment by MikeSilberbauer** [comment]
Sorry, I gave too much information - I wanted to inform you that the Metadata file at http://www.riv.co.za/wv/data/Silberbauer-2020.yaml now contains the Software Heritage ID for the Pascal program.
The file at https://github.com/ReScience/submissions/files/4832693/ReScience_C_Silberbauer_2020.7.pdf contains the edits that I made in response to your review. 
I am not sure what the next step in the process is.

**Comment by rougier** [comment]
Now I get it! I'm satistified with the correction and figure 5 is just fantastic ! @oliviaguest I'm satisfied with the correction and  recommend acceptation.

Minor typo at the top abot the title, it should read "Reproduction in Zoology" (Zoology?)

**Comment by MikeSilberbauer** [comment]
[ReScience_C_Silberbauer_2020 (9).pdf](https://github.com/ReScience/submissions/files/4913362/ReScience_C_Silberbauer_2020.9.pdf)

Thank you!

I have fixed a problem with the last pages in the Appendix, which had lost subsubheading floating around.

I am probably being a bit dim, but I have not been able to work out what part of the submission you are referring to regarding the "minor typo at the top" about the title. If this helps: the subject area is "Aquatic science" or "Limnology" or even "Water chemistry". So "Reproduction in Limnology". (?)

**Comment by rougier** [comment]
I meant the text just above the title that is generated automatically from the metadata. The type of work is "Reproduction" and for the domain you indicated "Aquatic Science" such that the heading should red "Reproduction in Aquatic Science"

**Comment by MikeSilberbauer** [comment]
[ReScience_C_Silberbauer_2020 (10).pdf](https://github.com/ReScience/submissions/files/4949069/ReScience_C_Silberbauer_2020.10.pdf)
OK, thank you for the explanation, now I understand. I had not entered all the required information into metadata.tex before compiling in Overleaf.

**Comment by rougier** [comment]
@oliviaguest I think we're good for publication!

**Comment by oliviaguest** [comment]
Fantastic!

**Comment by oliviaguest** [comment]
@rougier naive question I should know the answer to but are references like [1] meant to be clickable and then scroll to 1 in the reference list in the PDF?

**Comment by MikeSilberbauer** [comment]
The references probably are meant to work like that, but I couldn't find out how to get the bibliography to work automatically in the Overleaf editor, so I must confess that I hard-coded it. I can try to figure it out or manually insert cross-references if necessary.

**Comment by oliviaguest** [comment]
@rougier do you know how to address this? I think maybe giving a clear way to do it for overleaf might be a bit too much but surely a way to do it step-by-step is somewhere already, right? In the README? Have you tried that @MikeSilberbauer?

**Comment by MikeSilberbauer** [comment]
I will read the Overleaf instructions again and try to find out what I am missing.

**Comment by MikeSilberbauer** [comment]
It was as simple as putting this in the right place:
\addbibresource{bibliography.bib}

[ReScience_C_Silberbauer_2020 (11).pdf](https://github.com/ReScience/submissions/files/4970558/ReScience_C_Silberbauer_2020.11.pdf)

**Comment by oliviaguest** [comment]
@MikeSilberbauer in order to run:

```
./process.py --sandbox --metadata metadata.yaml --pdf article.pdf
```

I need your metadata and the ability to compile the PDF — can you link me to the repo for your article, please? Alternatively, you can do all the steps yourself if you prefer and just involve me at the very end, I think? Is that right, @rougier? See: https://github.com/ReScience/articles/blob/master/README.md

**Comment by MikeSilberbauer** [comment]
@oliviaguest 
All the files for compiling the PDF are on Overleaf at this link:
https://www.overleaf.com/5771591391kxmhgdrygdzw
I hope that helps?

**Comment by oliviaguest** [comment]
Fab — thank you. I might push some minor changes and recompile, if that's OK. ☺️

**Comment by MikeSilberbauer** [comment]
Yes, please do: I am very much a beginner with LaTeX and Github.

**Comment by oliviaguest** [comment]
Yeah, don't worry I'm struggling just as much with the new way of doing things! 🤣

**Comment by oliviaguest** [comment]
@rougier I am trying to compile the PDF, but get: 

```
(base) MBpro:5e9eb999ec43100001b5cbca olivia$ make
./yaml-to-latex.py -i metadata.yaml -o metadata.tex
make: ./yaml-to-latex.py: Permission denied
make: *** [metadata.tex] Error 1
```

What am I doing wrong?

**Comment by rougier** [comment]
`chmod 755 yaml-to-latex.py` maybe?

**Comment by oliviaguest** [comment]
Oh, ha! Thanks, @rougier. What do we put for the 2nd reviewer? Leave it blank, right?

**Comment by rougier** [comment]
Yes. If blank, it should not appear in the PDF.

**Comment by oliviaguest** [comment]
Is this OK, @rougier?
![3923a589-3695-436e-8d83-a4bf7f379cf3](https://user-images.githubusercontent.com/5082092/88647672-e90cec00-d0ce-11ea-91e4-97f2e75aed04.png)

**Comment by oliviaguest** [comment]
Also hmm... the references are not looking correct either, right? 🤔

**Comment by rougier** [comment]
References are wrong yes. And why do I get a n°2. Is my name listed as first or second reviewer in the metadata ?

**Comment by MikeSilberbauer** [comment]
If I understood what I was doing, the files that I edited in Overleaf have just been updated in https://github.com/MikeSilberbauer/Silberbauer-2020, in case that helps.

**Comment by oliviaguest** [comment]
@rougier @MikeSilberbauer Thank you both for the feedback/comments. Lemme try again...!

**Comment by oliviaguest** [comment]
@MikeSilberbauer can you make the repo you linked to public, please? The link 404s, so I assume it's private?

**Comment by MikeSilberbauer** [comment]
@oliviaguest - it should be accessible to you now, if I have read the manual correctly...

**Comment by oliviaguest** [comment]
@MikeSilberbauer can you add @rougier too, please? I have pushed these two files and I want to make sure they look right as the reviewer issue is not resolved @rougier. Do these files look right?

https://github.com/MikeSilberbauer/Silberbauer-2020/blob/master/article.pdf

https://github.com/MikeSilberbauer/Silberbauer-2020/blob/master/metadata.yaml

**Comment by MikeSilberbauer** [comment]
OK, @rougier is added

**Comment by oliviaguest** [comment]
@MikeSilberbauer please don't feel this is your fault BTW — I suck at this! 🤣

**Comment by MikeSilberbauer** [comment]
Not at all - I appreciate all the assistance. I used to wonder what Github was all about, so having to jump in at the deep end is one way of learning.

**Comment by rougier** [comment]
* In the metadata, type should be Reproduction (not Replication, the commented list is actually missing "* Reproduction")
* Title should be "[Rp] Geographical Trends in the Water Chemistry of Wetlands in the South-Western Cape Province, South Africa." (not "Re ReScience challenge: resurrecting a 30-year-old Turbo Pascal program for generating Maucha ionic diagrams")
* For the reviewer 2, something wrong. Maybe @MikeSilberbauer needs to use the latest ReScience template. Have a look at at this [entry](https://zenodo.org/record/3886628/files/article.pdf) for what it is supposed to look like

**Comment by MikeSilberbauer** [comment]
@rougier Your article is most fascinating, and I was impressed that your program could be loaded onto the online Apple emulator. The scanned pages from the original computer magazine remind me of the publication style of the day - I laboriously typed a couple of Fortran 7 program listings from journals onto the old corporate mainframe computer to see what they could do. There was an Apple ][ that acted as a process controller for a scintillation counter, but we junior staff were not allowed near it.

Regarding the version of the template that I used, could you please remind me of the correct link so that I can make sure?

**Comment by rougier** [comment]
Thanks :) Here is the link, hopefully, the "Reviewer 2" will disappear. Else, I'll need to investigate here the "2" comes from.

**Comment by MikeSilberbauer** [comment]
@oliviaguest - I have modified metadata.yaml and manually edited metadata.txt on https://github.com/MikeSilberbauer/Silberbauer-2020/:
https://github.com/MikeSilberbauer/Silberbauer-2020/blob/master/metadata.tex
In Overleaf, the notes in the left margin of the PDF file look better - I have placed a copy of the output from Overleaf here:
https://github.com/MikeSilberbauer/Silberbauer-2020/blob/master/ReScience_C_Silberbauer_2020.pdf

**Comment by rougier** [comment]
Looks good to me no. @oliviaguest will give you issue number and DOI.

**Comment by oliviaguest** [comment]
The info is there, it just needs to be compiled. However we go back to the [same problem as above from two weeks ago](https://github.com/ReScience/submissions/issues/25#issuecomment-664923332) when I compile the PDF. Can perhaps @rougier or @MikeSilberbauer compile it and see if they also have this problem? Maybe I'm looking at the wrong PDF, but when I run make this is the output: https://github.com/MikeSilberbauer/Silberbauer-2020/blob/master/article.pdf

Am I missing something? Ideas?

**Comment by MikeSilberbauer** [comment]
@oliviaguest: I think I forgot to make the change to the title in the metadata.yaml file - I manually changed it in metadata.tex, which gets overwritten, if I understand the process correctly. I have now made the change to the title and removed the blank second reviewer from metadata.yaml. Could you please check if the compile works now?

**Comment by oliviaguest** [comment]
Hmm... @MikeSilberbauer I get this currently, sadly: 

```
(base) MBpro:Silberbauer-2020 olivia$ make
./yaml-to-latex.py -i metadata.yaml -o metadata.tex
/Users/olivia/Code/Silberbauer-2020/article.py:170: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  document = yaml.load(data)
Traceback (most recent call last):
  File "./yaml-to-latex.py", line 88, in <module>
    content = generate_latex_metadata(filename_in, article)
  File "./yaml-to-latex.py", line 49, in generate_latex_metadata
    "".format(filename=filename, _=article, abstract=abstract))
IndexError: list index out of range
make: *** [metadata.tex] Error 1
```

I'll investigate.

**Comment by oliviaguest** [comment]
@rougier so [this change](https://github.com/MikeSilberbauer/Silberbauer-2020/commit/12a750eb3d3e44bf20d1ee878d4dd89f198bd3a8) breaks everything see comment above.

**Comment by rougier** [comment]
You  have to leave the second reviewer empty indeed. If things are borken, maybe the best is to change metadata.tex manually and then you're done.

**Comment by MikeSilberbauer** [comment]
@oliviaguest @rougier I have used the file renamed metadata_manual.tex in https://github.com/MikeSilberbauer/Silberbauer-2020 to generate ReScience_C_Silberbauer_2020.pdf in the same repository (using Overleaf).

**Comment by oliviaguest** [comment]
Ah, great, thanks @MikeSilberbauer. I noticed a typo: `(see Appendix, Figure 3.` Can you fix that and/or explain how you compiled it, please, so I can compile and submit it. 😊

**Comment by MikeSilberbauer** [comment]
@oliviaguest Thanks for picking up that missing ")". I have corrected it and recompiled in Overleaf. The new PDF file is named ReScience_C_Silberbauer_2020.pdf in https://github.com/MikeSilberbauer/Silberbauer-2020

**Comment by oliviaguest** [comment]
Great! @MikeSilberbauer can you either explain how you compiled it and/or fix the date to be today's date for publication? And then I'll try and get it done today, otherwise you might need to put tomorrow's date. 😅

**Comment by MikeSilberbauer** [comment]
@oliviaguest No stress: I have changed the date to **tomorrow** in the metadata.tex file in the Overleaf version of the publication, generated the PDF file and copied it as ReScience_C_Silberbauer_2020.pdf in https://github.com/MikeSilberbauer/Silberbauer-2020

**Comment by oliviaguest** [comment]
@rougier what am I doing wrong now? 😆 😭 

```
(base) MBpro:articles olivia$ ./publish.py --sandbox --metadata ../Silberbauer-2020/metadata.yaml --pdf ../Silberbauer-2020/ReScience_C_Silberbauer_2020.pdf 
Uploading content to Zenodo... Traceback (most recent call last):
  File "./publish.py", line 204, in <module>
    upload_content(server, token, article_id, article_file)
  File "./publish.py", line 25, in upload_content
    raise IOError("%s: " % response.status_code + response.json()["message"])
OSError: 404: PID does not exist.
```

On Linux too: 

```
(base) olivia@olivia-HP-Z2-Tower-G4-Workstation:~/Code/articles$ ./publish.py --sandbox --metadata ../Silberbauer-2020/metadata.yaml --pdf ../Silberbauer-2020/ReScience_C_Silberbauer_2020.pdf 
Uploading content to Zenodo... Traceback (most recent call last):
  File "./publish.py", line 204, in <module>
    upload_content(server, token, article_id, article_file)
  File "./publish.py", line 25, in upload_content
    raise IOError("%s: " % response.status_code + response.json()["message"])
OSError: 404: PID does not exist.
```

**Comment by rougier** [comment]
It depends when you reserved the DOI on Zenodo. (sandbox). You need first to reserve the DOI sung the process, update the PDF and then publish it.

**Comment by rougier** [comment]
Best is to have the latex sources ready and change the DOI once you get it, recompile and than publish it. Note that you'll also need to update the `metadata.yaml` (even if you don't use it to re-generate the `metadta.tex`)

**Comment by oliviaguest** [comment]
How do you do generate the PDF without using `metadata.yaml`?

**Comment by oliviaguest** [comment]
Also I'm confused @rougier... do you mean reserve another DOI? Because I already did that once.

**Comment by rougier** [comment]
You edit the `metadata.yaml` with the proper DOI and you also edit the `metadata.tex` manually (just in cas you have problem going from yaml to tex). For the SANDBOX Zenodo, you can request another DOI and try with this one. These DOI are erased every day I think.

**Comment by otizonaizit** [comment]
@oliviaguest : if you want we could try to look at the whole process together and see what does and does not work? I have not experienced so many issues as you seem to have while using the new publication process, so there may be some implicit step which is not listed in the docs? Or there may be some implicit software dependencies which are by chance present on my sytem and absent from yours? In any case, publishing should not be such a pain, so I think it's worth it to try to improve the process/the docs.

**Comment by oliviaguest** [comment]
Hmmm, I am sure @rougier I reserved a real DOI without sandbox.

**Comment by oliviaguest** [comment]
@otizonaizit sure, over video? when is good for you? I'm on UTC + 3.

**Comment by oliviaguest** [comment]
@otizonaizit sadly I still have the same problem after requesting a new DOI! See repo: https://github.com/oliviaguest/Silberbauer-2020/

```bash
(base) MBpro:articles olivia$ ./process.py --sandbox --metadata ../Silberbauer-2020/metadata.yaml --pdf ../Silberbauer-2020/article.pdf
Article ID:  663987
Article DOI: 10.5072/zenodo.663987
Article URL: https://sandbox.zenodo.org/record/663987/files/article.pdf
(base) MBpro:articles olivia$ ./process.py --zenodo --metadata ../Silberbauer-2020/metadata.yaml --pdf ../Silberbauer-2020/article.pdf
Article ID:  3996198
Article DOI: 10.5281/zenodo.3996198
Article URL: https://zenodo.org/record/3996198/files/article.pdf
(base) MBpro:articles olivia$ ./publish.py --sandbox --metadata ../Silberbauer-2020/metadata.yaml --pdf ../Silberbauer-2020/article.pdf
Uploading content to Zenodo... Traceback (most recent call last):
  File "./publish.py", line 204, in <module>
    upload_content(server, token, article_id, article_file)
  File "./publish.py", line 25, in upload_content
    raise IOError("%s: " % response.status_code + response.json()["message"])
OSError: 404: PID does not exist.
```

**Comment by otizonaizit** [comment]
@oliviaguest : what happens if you try the `publish.py` script directly against the official zenoo and not against the sandbox? The zenodo sandbox has not work reliably for me either in the past. Given that the PDF is now correct and the metadata also, I think there's no risk in trying to do the publication on zenodo directly.
If that fails too, maybe we should have another video-chat and see if this is indeed a mac vs Linux issue...

**Comment by oliviaguest** [comment]
OK, I think that worked!
